docs: ibis-analytics blog post (#7990)
ibis-project/ibis: docs/posts/ibis-analytics/index.qmd

---
title: Modern, hybrid, open analytics
author: "Cody"
date: "2024-01-16"
image: "thumbnail.png"
categories:
    - blog
    - duckdb
    - bigquery
    - case study
---

## Introduction

As a Python data user, I've wanted a more modular, composable, and scalable
ecosystem. *I think it's here*. Wes McKinney released pandas c. 2009 to bring
dataframes into Python and it became one of the most used software packages. It
was built when small data was smaller and though still an excellent dataframe
library today has some downsides [Wes wrote about in his "Apache Arrow and the
'10 things I hate about pandas'" blog

> post](https://wesmckinney.com/blog/apache-arrow-pandas-internals). Wes created

Reviewer: This sentence is a bit too long, and I'm losing the train of thought. Not sure what's the bet way, but maybe something like: 
"I was built when small data was smaller and though still an excellent dataframe
library today. But it has some downsides, [Wes wrote about in his "Apache Arrow and the
'10 things I hate about pandas'" blog
post](https://wesmckinney.com/blog/apache-arrow-pandas-internals)."

---
title: Modern, hybrid, open analytics
author: "Cody"
date: "2024-01-20"
image: "thumbnail.png"
categories:
    - blog
    - duckdb
    - bigquery
    - case study
---

## Introduction

As a Python data user, I've wanted a more modular, composable, and scalable
ecosystem. *I think it's here*. Wes McKinney released pandas c. 2009 to bring

> dataframes into Python and it became one of the most used software packages. It

Reviewer 2: ```suggestion
the power of SQL to Python via the DataFrame and it became one of the most used software packages. It
```

I think this may appeal to a slightly more general audience.

Author: this is on pandas? it didn't use/bring the power of SQL?

Reviewer 2: Literally no, but it brought the concepts (and the conceptual power) of SQL to the forefront of the Python ecosystem.

was built when small data was smaller and though still an excellent dataframe
library today has some downsides [Wes wrote about in his "Apache Arrow and the
'10 things I hate about pandas'" blog
post](https://wesmckinney.com/blog/apache-arrow-pandas-internals). Wes created
Ibis (and co-created Apache Arrow) to address these issues and more with a
different approach than any other Python dataframe library by decoupling the
dataframe API from the backend implementation. This allows Ibis to support 20+
backends today, including pandas and DuckDB and Polars and Postgres and BigQuery

> and Snowflake and PySpark and over a dozen more.

Reviewer 2: I'm not sure we need all the "and"s :)

Author: changed to commas

I've experienced many pains of the ecosystem myself. I've downloaded a 1 GB CSV
from a database GUI and used pandas to munge it because I couldn't figure it out
in SQL. When building larger ML demos that included data preprocessing on GBs, I

> hit OOM errors and learned about how much data expands in memory and why to
> store files in compressed Parquet. I learned about partitioning strategies and

Reviewer 2: ```suggestion
hit OOM errors and learned about how much data expands in memory and why
storing data in Parquet files might be a better option for my use cases. I learned about partitioning strategies and
```

> the joy of subtle differences and incompatibilities across data system. I

Reviewer 2: ```suggestion
the joy of subtle differences and incompatibilities across data systems. I
```

started using PySpark and Dask to overcome hurdles. I slowly learned about
databases. Fortunately while I've been learning, others ([including many Ibis
contributors](https://github.com/ibis-project/ibis/contributors)) have been

> building a modern, open-source Python data ecosystem! Now I can just use what
> they've built.

Reviewer 2: ```suggestion
building a modern, open-source Python data ecosystem! Now I can use what
they've built to create a modern analytics application.
```

In this blog, we'll look at how to build an end-to-end analytics project using

> Ibis as the frontend for data in Python. Ibis plays a large part, but is only

Reviewer 2: ```suggestion
Ibis as the frontend for data in Python. Ibis plays a key role, but is only
```

one component in any real data project.

> We'll combine a variety of open-source tools and freemium services including:

Reviewer 2: It could be useful to partition this into two groups: software components and data, since the data are software-related.

Author: I like the big list

- [Ibis](https://ibis-project.org) (dataframe library)
- [DuckDB](https://duckdb.org) (database and query engine)
- [Dagster](https://dagster.io) (orchestration library)
- [Plotly](https://plotly.com/python) (visualization library)
- [Streamlit](https://streamlit.io) (dashboard library)
- [justfile](https://just.systems) (command runner)
- [TOML](https://toml.io) (configuration)
- [GitHub](https://github.com) (source control, CI/CD)
- [Azure VM](https://azure.microsoft.com/products/virtual-machines) (self-hosted runner)
- [GitHub web APIs](https://docs.github.com/en/graphql) (source data)
- [Google BigQuery](https://cloud.google.com/free/docs/free-cloud-features#bigquery) (source data)
- [Zulip](https://zulip.com) (source data)
- [MotherDuck](https://motherduck.com) (cloud service for DuckDB)
- [Streamlit Community Cloud](https://docs.streamlit.io/streamlit-community-cloud) (cloud service for Streamlit)

to build an open-source, modular, composable, scalable, automated, hybrid-cloud,
end-to-end analytics project processing a few million rows of data that provides

> real business value at little to no cost.

Reviewer 2: ```suggestion
real business value.
```

Probably overpromising a bit here. I think "real business value" is enough justification.

## What's the business value?

Today, Ibis is a thriving open-source project primarily backed by Voltron Data
with contributors at Google, Claypot AI, Singlestore, Exasol, RisingWave,
Starburst Data, and anyone who [submits a PR and makes a
commit](https://github.com/ibis-project/ibis/contributors). It aims to be a
standard frontend for data in Python that scales to your backend. To understand
the project's health, we want to track some key adoption metrics. There was
already a dashboard internally at Voltron Data, but it was written in R and I
don't know R. I do know Ibis and a few other OSS tools, and saw this as a cool

> opportunity to try out a few more.

Reviewer 2: ```suggestion
opportunity to try out a few more and showcase Ibis use in an application.
```

Let's get started!

## Finished product

The final result is a MotherDuck database powering a dashboard deployed to
Streamlit Community Cloud.

::: {.callout-warning}

> While I consider this "production" it does go occasionally, usually because of a

Reviewer 2: ```suggestion
While I consider this "production" it does go down occasionally, usually because of a
```

code](https://github.com/ibis-project/ibis-analytics). A point-in-time
description and snapshots of the code are below.

### Goals

I had the following goals in mind:

- **Modular:** while this project uses Dagster, it should be easily swappable for other orchestration tools like Airflow or Prefect or Kedro. While this project uses DuckDB, it should be easily swappable for Polars or Clickhouse or any other engine. While this project uses Streamlit, it should be swappable for Quarto dashboards...
- **Composable:** the project should be easy to understand and extend. It should be easy to add new data sources, new data transformations, new metrics, and new visualizations.
- **Automated:** the project should be easy to run locally and in production. It should be easy to run the entire pipeline or just a single step. It should be easy to run the pipeline on a schedule or on a pull request.
- **Iterative:** the project should be easy to iterate on. It should be easy to explore the data, transform it, and visualize it. It should be easy to switch between local and production data sources.

Our open-source tools and freemium services allow us to accomplish these goals

> at little to no cost.

Reviewer 2: ```suggestion
at little to no cost.
```

Not sure what to say instead here but if we can avoid general statements about how much something costs that'd be best I think, to avoid giving the impression that anything you'll build is going to be low-cost.

### Overview

The architecture consists of Python code. Steps in the overall data pipeline are
captured in a `justfile`, which is a Makefile-like command runner. This allows
us to run the pipeline locally or in automated CI/CD workflows. The data
ingestion is a Python script that makes raw `request` calls to GitHub web APIs,
uses the Zulip Python client for Zulip web APIs, and Ibis + SQL to extract PyPI
data from a public Google BigQuery project.

The data is then extracted from its ingested file formats (JSON and Parquet),
transformed with Ibis using the default DuckDB backend, and loaded into final
DuckDB database files. Further postprocessing is done to combine these into a
single DuckDB database file for convenience, though this step isn't necessary.

> The tables are then copied using Ibis to connect to MotherDuck, a serverless
> cloud service for DuckDB. This allows us to use MotherDuck's serverless servers

Reviewer 2: ```suggestion
The tables are then copied using Ibis to connect to MotherDuck, a cloud service for DuckDB. MotherDuck allows us
```

Very strange to read "serverless servers". Are they servers or not? 😅

    )
    downloads = downloads.mutate(
        major_minor_patch=f.clean_version(downloads["version"], patch=True),
        major_minor=f.clean_version(downloads["version"], patch=False).cast("float"),
    )
    downloads = downloads.rename(
        {
            "version_raw": "version",
            "version": "major_minor_patch",
        }
    )
    downloads = (
        downloads.group_by(
            [
                ibis._.timestamp.truncate("D").name("timestamp"),
                ibis._.country_code,
                ibis._.version,
                ibis._.python,
                ibis._.system["name"].name("system"),
            ]
        )
        .agg(
            ibis._.count().name("downloads"),
        )
        .order_by(ibis._.timestamp.desc())
        .mutate(
            ibis._.downloads.sum()
            .over(
                rows=(0, None),
                group_by=["country_code", "version", "python", "system"],
                order_by=ibis._.timestamp.desc(),
            )
            .name("total_downloads")
        )
        .order_by(ibis._.timestamp.desc())
    )
    downloads = downloads.mutate(ibis._["python"].fillna("").name("python_full"))

>     downloads = downloads.mutate(

Reviewer 2: All this reassignment to the same variable suggests that perhaps you should chain methods together.

You don't lose any debuggability, since you can successively comment parts of the chain in or out.

Author: 🤷 it's how the code is currently written and works

        f.clean_version(downloads["python_full"], patch=False).name("python")
    )
    return downloads
```

Like the extract stage, the output of the transform stage is written to DuckDB
database files for use downstream in the DAG.

### Load

Honestly this step isn't doing anything. It could use Ibis to directly upload
the tables to MotherDuck, but as implemented these assets are just passing
through the transformed data. This is a bit wasteful but allows for a
three-stage DAG that conforms to the ETL paradigm.

```{.python}

> # imports

Reviewer 2: Can you remove these `# imports` comments? They aren't providing much value since the next thing that follows is a bunch of import statements :)

Author: 90% of my code comments aren't providing much value...removing these ones though
